inference-engine/src/inference_engine/include/openvino/core/tensor.hpp

```diff
@@ -0,0 +1,155 @@
+
+namespace ov {
+namespace runtime {
+class InferRequest;
+class RemoteContext;
+class VariableState;
+}  // namespace runtime
+
+/**
+ * @brief Memory access and interpretation API
+ *
+ * It can throw exceptions safely for the application, where it is properly handled.
+ */
+class INFERENCE_ENGINE_API_CLASS(Tensor) {
+protected:
+    std::shared_ptr<void> _so;        //!< Refernce to dynamicly loaded library
+    std::shared_ptr<ie::Blob> _impl;  //!< Shared pointer to internal tensor representation
+
+    /**
+     * @brief Constructs Tensor from the initialized std::shared_ptr
+     * @param so Plugin to use. This is required to ensure that Tensor can work properly even if plugin object is
+     * destroyed.
+     * @param impl Initialized shared pointer
+     */
+    Tensor(const std::shared_ptr<void>& so, const std::shared_ptr<ie::Blob>& impl);
+
+    friend class ov::runtime::InferRequest;
+    friend class ov::runtime::RemoteContext;
+    friend class ov::runtime::VariableState;
+
+public:
+    /**
+     * @brief Default constructor
+     */
+    Tensor() = default;
+
+    /**
+     * @brief Constructs Tensor using element type and shape. Allocate internal host storage using default allocator
+     * @param type Tensor element type
+     * @param shape Tensor shape
+     * @param allocator allocates memory for internal tensor storage
+     */
+    Tensor(const element::Type type, const Shape& shape, const Allocator& allocator = {});
+
+    /**
+     * @brief Constructs Tensor using element type and shape. Wraps allocated host memory.
+     * @note Dose not own the memory
+     * @param type Tensor element type
+     * @param shape Tensor shape
+     * @param ptr Pointer to allocate memory
+     * @param size Optional size of allocated host memory in elements. If it is not used the size of memory supposed to
+     * be not less then ov::shape_size(shape) * type.size() in bytes.
+     * @param strides Optional element strides paramters. It is supposed that strides is equal to shape if it is not
+     * used
+     */
+    Tensor(const element::Type type, const Shape& shape, void* ptr, const size_t size = 0, const Strides& strides = {});
+
+    /**
+     * @brief constructs region of interest (ROI) tensor form other tensor.
+     * @note Dose not own the memory
+     * @param other original tensor
+     * @param begin start coordinate of ROI object inside of the original object.
+     * @param end end coordinate of ROI object inside of the original object.
+     */
+    Tensor(const Tensor& other, const Coordinate& begin, const Coordinate& end);
+
+    /**
+     * @brief Set new shape for tensor, deallocate/allocate if new total size is bigger than previous one.
+     *
+     * @param shape new shape
+     */
+    void set_shape(const ov::Shape& shape);
+
+    /**
+     * @return tensor element type
+     */
+    element::Type get_element_type() const;
+
+    /**
+     * @return tensor element type
+     */
+    Shape get_shape() const;
```

Comment:
Just curious, is it possible to return such objects by constant reference instead of by value? Just copies of such objects might lead to heap memory allocation and extra memory copies, which might affect performance if the method is called quite often.